core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -1334,6 +1334,37 @@ class UnifiedLog(@volatile var logStartOffset: Long,
         } else {
           Some(new TimestampAndOffset(RecordBatch.NO_TIMESTAMP, -1L, Optional.of(-1)))
         }
+      } else if (targetTimestamp == ListOffsetsRequest.EARLIEST_PENDING_UPLOAD_OFFSET_TIMESTAMP) {
+        if (remoteLogEnabled()) {
+          val curHighestRemoteOffset = highestOffsetInRemoteStorage()
+
+          if (curHighestRemoteOffset == -1) {
+            if (localLogStartOffset() == logStartOffset) {
+              // No segments have been uploaded yet
+              fetchOffsetByTimestamp(ListOffsetsRequest.EARLIEST_TIMESTAMP, remoteLogManager)
+            } else {
+              // Leader currently does not know about the already uploaded segments
+              Some(new TimestampAndOffset(RecordBatch.NO_TIMESTAMP, -1L, Optional.of(-1)))
+            }
+          } else {
+            val earliestPendingUploadOffset = math.max(curHighestRemoteOffset + 1, logStartOffset)
+
+            val epochResult: Optional[Integer] =
+              if (leaderEpochCache.isDefined) {
+                val epochOpt = leaderEpochCache.get.epochForOffset(earliestPendingUploadOffset)
+                if (epochOpt.isPresent) {
+                  Optional.of(epochOpt.getAsInt)
+                } else {
+                  Optional.empty()
+                }
+              } else {
+                Optional.empty()
+              }
+            Some(new TimestampAndOffset(RecordBatch.NO_TIMESTAMP, earliestPendingUploadOffset, epochResult))
```

Comment:
```suggestion
            val epochResult = leaderEpochCache.flatMap(_.epochForOffset(earliestPendingUploadOffset).asScala)
              .asJava.asInstanceOf[Optional[Integer]]
            Some(new TimestampAndOffset(RecordBatch.NO_TIMESTAMP, earliestPendingUploadOffset, epochResult))
```

Reply:
Done. Thanks for the suggestion.

Reply:
There is other places in the same method can benefit from this refactor as well like  line 1319 and 1298

Reply:
Made the requested changes. Also changed for LATEST_TIMESTAMP